### DISCUSSION TITLE
Tools that relies on the path to the `zstd` library break after updating the library

### DISCUSSION THREAD

**Discussion by voodoos** [Homebrew/discussions#4682] [Everyday usage]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.1.1
ORIGIN: https://github.com/Homebrew/brew
HEAD: 3b3300546b5a4e40b74f4ee33cf225cca280defe
Last commit: 7 days ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: e9f831371f60b4035c20c9c953f6d8c23a0b3f8e
Core tap last commit: 5 months ago
Core tap branch: master
Core tap JSON: 25 Jul 13:43 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 14.0.3 build 1403
Git: 2.39.2 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 7.87.0 => /usr/bin/curl
macOS: 13.3.1-arm64
CLT: 14.3.0.0.1.1679647830
Xcode: 14.3.1
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Your system is ready to brew.
```

### Description of issue

I think this issue is specific to the `zstd` formula, I can move that post somewhere else if there is a place for that.

I am using a tool, the OCaml compiler, that links programs with the `zstd` library. When the compiler is configured and built, it uses the lib's `.pc` file to get the path to the library. When linking an ocaml program with the compiler, this path is passed to `ld -L...`.

The issue is that, when `zstd` is installed on a mac with `brew`, the path in the `.pc` file points to a specific version of the `zstd` instead of the more generic `/opt/homebrew/opt/`. This means that when the library is updated the link breaks and re-configuring (and rebuilding) the compiler is necessary to get the path to the new version of the library. This seems sub-optimal given the existence of a symlink to the library in the `/opt/homebrew/lib` that points to the latest version. Having this path in the `.pc` file would solve the issue.

Sorry if my explanations are a bit blurry, I am not an expert in these matters, but I can provide more context / debug information if required. I can also move my post to a more appropriate place if that exists.

Original issue: https://github.com/ocaml/ocaml/issues/12431

The error shown when running a compiler built with an outdated `zstd` lib:
```
$ ocamlopt.opt main.ml
ld: warning: directory not found for option '-L/opt/homebrew/Cellar/zstd/1.5.2/lib'
ld: library not found for -lzstd
clang: error: linker command failed with exit code 1 (use -v to see invocation)
File "caml_startup", line 1:
Error: Error during linking (exit code 1)
```

**Comment by osalbahr** [comment]
Could you show where you set `'-L/opt/homebrew/Cellar/zstd/1.5.2/lib'`? I might be able to think of a workaround using `brew` commands to avoid hardcoding the version (see [`brew --cellar [formula]`](https://docs.brew.sh/Manpage#--cellar-formula-).

Edit: I posted a couple workarounds in the upstream

- https://github.com/ocaml/ocaml/issues/12431#issuecomment-1658924961

---

~~A more permanent fix is to have Homebrew have a shared `/lib` directory where shared libraries are dynamically symlinked upon upgrading (similarly to `"$(brew --prefix)/bin"`). Not sure if that would cause naming collisions. Or, maybe a symlinked `zstd` would suffice (to have a path that does not change upon `brew upgrade`)~~

Edit `$(brew --prefix)/lib` is already a thing.

**Reply by osalbahr** [reply]
Does [brew install ocaml](https://github.com/ocaml/ocaml/issues/12431#issuecomment-1658957253) fit your use case?

**Comment by voodoos** [comment]
Thanks for your reply @osalbahr !
I answered in the ocaml's issue. I guess it's best to keep the discussion there and then report here :-)

**Comment by voodoos** [comment]
A patch was made to write the generic path to the lib in the pc file: https://github.com/Homebrew/homebrew-core/pull/138187

This should fix the issue we were having, thank you !

**Reply by osalbahr** [reply]
You’re welcome!
